pkg/storage/odf_test.go

```diff
@@ -0,0 +1,351 @@
+			client:              true,
+		},
+		{
+			name:                "odftest",
+			namespace:           "openshift-storage",
+			addToRuntimeObjects: false,
+			expectedError:       fmt.Errorf("storageSystem object odftest does not exist in namespace openshift-storage"),
+			client:              true,
+		},
+		{
+			name:                "odftest",
+			namespace:           "openshift-storage",
+			addToRuntimeObjects: true,
+			expectedError:       fmt.Errorf("storageSystem 'apiClient' cannot be empty"),
+			client:              false,
+		},
+	}
+
+	for _, testCase := range testCases {
+		// Pre-populate the runtime objects
+		var runtimeObjects []runtime.Object
+
+		var testSettings *clients.Settings
+
+		testStorageSystem := generateStorageSystem(testCase.name, testCase.namespace)
+
+		if testCase.addToRuntimeObjects {
+			runtimeObjects = append(runtimeObjects, testStorageSystem)
+		}
+
+		if testCase.client {
+			testSettings = clients.GetTestClients(clients.TestClientParams{
+				K8sMockObjects: runtimeObjects,
+			})
+		}
+
+		builderResult, err := PullStorageSystem(testSettings, testCase.name, testCase.namespace)
+		assert.Equal(t, testCase.expectedError, err)
+
+		if testCase.expectedError == nil {
+			assert.Equal(t, testStorageSystem.Name, builderResult.Object.Name)
+			assert.Equal(t, testStorageSystem.Namespace, builderResult.Object.Namespace)
+		}
+	}
+}
+
+func TestNewStorageSystemBuilder(t *testing.T) {
+	testCases := []struct {
+		name          string
+		namespace     string
+		expectedError string
+		client        bool
+	}{
+		{
+			name:          defaultStorageSystemName,
+			namespace:     defaultStorageSystemNamespace,
+			expectedError: "",
+			client:        true,
+		},
+		{
+			name:          "",
+			namespace:     defaultStorageSystemNamespace,
+			expectedError: "storageSystem 'name' cannot be empty",
+			client:        true,
+		},
+		{
+			name:          defaultStorageSystemName,
+			namespace:     "",
+			expectedError: "storageSystem 'nsname' cannot be empty",
+			client:        true,
+		},
+		{
+			name:          defaultStorageSystemName,
+			namespace:     defaultStorageSystemNamespace,
+			expectedError: "",
+			client:        false,
+		},
+	}
+
+	for _, testCase := range testCases {
+		var testSettings *clients.Settings
+
+		if testCase.client {
+			testSettings = clients.GetTestClients(clients.TestClientParams{})
+		}
+
+		testStorageSystemBuilder := NewStorageSystemBuilder(testSettings, testCase.name, testCase.namespace)
+
+		if testCase.expectedError == "" {
+			if testCase.client {
+				assert.Equal(t, testCase.name, testStorageSystemBuilder.Definition.Name)
+				assert.Equal(t, testCase.namespace, testStorageSystemBuilder.Definition.Namespace)
+			} else {
+				assert.Nil(t, testStorageSystemBuilder)
+			}
+		} else {
+			assert.Equal(t, testCase.expectedError, testStorageSystemBuilder.errorMsg)
+			assert.NotNil(t, testStorageSystemBuilder.Definition)
+		}
+	}
+}
+
```

Comment:
I noticed that Update function is missing. Can you add it?

Reply:
it is not relevant in this case. You cannot update. you can only delete and create new